core/core_to_map_std/src/unstable/services.rs

```diff
@@ -1,37 +1,91 @@
+#[derive(Debug, thiserror::Error)]
+pub enum PrepareServicesMapError {
+    #[error("Service is misconfigured:\n{}", ServiceMisconfiguredError::format_errors(.0.as_slice()))]
+    ServicesMisconfigured(Vec<ServiceMisconfiguredError>)
+}
+impl From<PrepareServicesMapError> for PerformException {
+    fn from(value: PrepareServicesMapError) -> Self {
+        PerformException {
+            error_code: PerformExceptionErrorCode::PrepareServicesMapError,
+            message: value.to_string(),
+        }
+    }
+}
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct ServiceMisconfiguredError {
+    pub id: String,
+    pub expected: String,
+}
+impl ServiceMisconfiguredError {
+    pub fn format_errors(errors: &[Self]) -> String {
+        let mut res = String::new();
+
+        for err in errors {
+            writeln!(
+                &mut res,
+                "Service {} is misconfigured. Expected {}",
+                err.id, err.expected
+            )
+            .unwrap();
+        }
+
+        res
+    }
+}
+
+pub fn prepare_services_map(provider_json: &ProviderJson, parameters: &MapValueObject) -> Result<MapValue, PrepareServicesMapError> {
     let mut services_map = MapValueObject::new();
+    let mut errors = Vec::new();
 
     for service in &provider_json.services {
-        let service_url = replace_parameters(service.base_url.clone(), parameters);
-        services_map.insert(service.id.to_string(), MapValue::String(service_url));
+        match replace_parameters(&service.base_url, parameters) {
+            Ok(service_url) => { services_map.insert(service.id.to_string(), MapValue::String(service_url)); }
+            Err(errs) => errors.extend(
+                errs.into_iter().map(|expected| ServiceMisconfiguredError {
+                    id: service.id.to_owned(),
+                    expected
+                })
+            )
+        }
     }
 
-    MapValue::Object(services_map)
+    if !errors.is_empty() {
+        Err(PrepareServicesMapError::ServicesMisconfigured(errors))
+    } else {
+        Ok(MapValue::Object(services_map))
+    }
 }
 
-fn replace_parameters(url: String, parameters: &MapValueObject) -> String {
+fn replace_parameters(url: &str, parameters: &MapValueObject) -> Result<String, Vec<String>> {
     let re = Regex::new(r"\{\s*([^}\s]*)\s*\}").unwrap();
 
-    let mut result_url = url.clone();
-
-    for cap in re.captures_iter(url.as_str()) {
+    let mut errors = Vec::new();
+    let result_url = re.replace_all(url, |cap: &Captures| {
         let param_name = &cap[1];
-        let param_value = parameters.get(param_name).unwrap();
 
-        match param_value {
-            MapValue::String(val) => {
-                result_url = result_url.as_str().replace(&cap[0], val);
+        match parameters.get(param_name) {
+            Some(MapValue::String(val)) => val,
+            _ => {
+                errors.push(format!("String parameter {}", param_name));
```

Comment:
Should error mention what is wrong?

```suggestion
                errors.push(format!("String parameter {} is missing", param_name));
```

Reply:
👍 

For posterity, the full error message (as reported by Python OneSDK):
```
ERROR: PrepareServicesMapError: Service is misconfigured:
Service default is misconfigured. Expected String parameter DOMAIN [is missing]
```

Reply:
Heh, but that gets caught by JSON validator, so it mostly likely will never be seen by the user:
```
INVALID INPUT: For Parameters following validation errors occured
  OneOfNotValid on path: , value: Object {"DOMAIN": Number(1), "PARAM": String("parameter_value")}
```